Vectorized `GetNonRandomizedHashCode`

Ruihan-Yin opened the pull request:
# Description

This PR vectorizes the hashing algorithm used in `GetNonRandomizedHashCode` and `GetNonRandomizedHashCodeIgnoreCase` with `Vector128` APIs. This is motivated by the fact that this method was spotted as the hot path in some first party workload.

In terms of the implementation, the incoming string input will be hashed on the vector path as long as the length is above the threshold. This leads to a 2-path hashing, meaning short strings (length under threshold) and long strings (length above threshold) will be hashed in different way, I'm not sure if this discrepancy matters, and the given structure might make it hard to scale up to V256 and V512.

we are open to discuss and improve the implementation.

Ruihan-Yin committed:
Improve the performance for GetNonRandomizedHashCode

Ruihan-Yin committed:
Improve the performance for GetNonRandomizedHashCode

Ruihan-Yin committed:
Ensure all the strings are hashed on the same path when SSE is available.

Ruihan-Yin committed:
bug fix attempt

Ruihan-Yin committed:
bug fix attempt

Ruihan-Yin committed:
bug fix attempt

Ruihan-Yin committed:
bug fix attempt

Ruihan-Yin commented:
Hi @EgorBo, could you help tag the person who can review this PR? I am not sure who I should include in the loop. Thanks

jkotas commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -815,34 +816,82 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
+                uint* ptr = (uint*) src;
                 int length = this.Length;
 
+                // We "normalize to lowercase" every char by ORing with 0x0020. This casts
```
This comment is misplaced. Copy&paste bug?

Ruihan-Yin replied:
Yes, will clean it up, thanks for pointing out.

jkotas commented:
> I'm not sure if this discrepancy matters

I think this discrepancy is a problem. It seems to violate rules for writing vectorized code in CoreLib: https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/vectors-and-intrinsics.md#crossgen2-adjustment-to-rules-for-systemprivatecorelibdll

jkotas commented:
Note that this algorithm is not set in stone. It should be fine to change the algorithm to one that is more vectorization friendly; however both the vectorized and non-vectorized paths need to produce same result.

EgorBo commented:
> Hi @EgorBo, could you help tag the person who can review this PR? I am not sure who I should include in the loop. Thanks

I presume the best people to review this are @GrabYourPitchforks and @stephentoub

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -815,34 +816,82 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
+                    {
+                        Vector128<uint> srcVec = Vector128.Load(ptr);
+                        length -= 8;
+                        hashVector = Vector128.Xor(Vector128.Add(hashVector, RotateLeft(hashVector, 5)), srcVec);
```
nit: we prefer operators over static methods, you can rewrite this into:
```cs
hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ srcVec;
```

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -1066,5 +1168,20 @@ private static CompareOptions GetCompareOptionsFromOrdinalStringComparison(Strin
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool VectorContainsNonAsciiChar(Vector128<ushort> utf16Vector)
```
such helpers already exists in Ascii.Utility.cs (VectorContainsNonAsciiChar)

Ruihan-Yin replied:
This method is private so might not be accessed here.

jkotas replied:
This is fine to make that method internal if needed.

Ruihan-Yin committed:
Try to make the vector path and scalar path output the same result.

Ruihan-Yin committed:
Remove unnecessary assertion

Ruihan-Yin committed:
Fix bad naming and infinite loop

Ruihan-Yin commented:
I tried to unroll the vector operation on the scalar path, now the scalar path and vector path should be able to converge to the same results. 

The tests failure should be due to the [GenerateCollidingStrings](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionary.Generic.Tests.cs#L69C37-L69C61) method is using the original hashing method and the new implementation will not collide on the pre-set collision point. Hence, to fix the failure, we probably need to study the new hash distribution and find the new colliding point.

Do we expect to fix the failures first, or we can review the implementation for now?

jkotas commented:
> Hence, to fix the failure, we probably need to study the new hash distribution and find the new colliding point.

Yes, this test needs to be fixed.

Could you please share before / after perf numbers for different string sizes?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -815,40 +816,107 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
                 int length = this.Length;
 
-                while (length > 2)
+                if(Vector128.IsHardwareAccelerated && length >= 2 * Vector128<ushort>.Count)
```
```suggestion
                if (Vector128.IsHardwareAccelerated && length >= 2 * Vector128<ushort>.Count)
```
Nit: The repo coding style is to use space between `if` and `(`. (I am not commenting on all places.)

jkotas commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -815,40 +816,107 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
+                {
+                    Vector128<uint> hashVector = Vector128.Create(hash1);
+
+                    while(length > 8)
```
```suggestion
                    while (length > 8)
```
Dtto for `while`

Ruihan-Yin commented:
> > Hence, to fix the failure, we probably need to study the new hash distribution and find the new colliding point.
> 
> Yes, this test needs to be fixed.
> 
> Could you please share before / after perf numbers for different string sizes?

Thanks for the reply, I will work on both shortly, I presume I could get some micro benchmark numbers on `ConcurrentDictionary` as examples.

jkotas commented:
> I could get some micro benchmark numbers on ConcurrentDictionary as examples, or do we want bare perf numbers on this method?

Both bare numbers on this method, and number for Dictionary or ConcurrentDictionary would be useful.

Ruihan-Yin committed:
Bug fix:
Update the hash collision generator in the test code.

Ruihan-Yin committed:
resolve comments

Ruihan-Yin committed:
make the expected collision point to be properly represented by int

Ruihan-Yin committed:
update test code, ensure to get the correct collision point

Ruihan-Yin committed:
bug fix

Ruihan-Yin committed:
bug fix

jkotas commented on src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs:
```diff
@@ -815,40 +817,107 @@ internal static int GetHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> value)
+                    uint hashed3 = hashVector.GetElement(2);
+                    uint hashed4 = hashVector.GetElement(3);
+
+                    while (length > 4)
```
I am not sure why you have changes this to `while` loop. I think `if` was perfectly fine here and more efficient too.

Ruihan-Yin replied:
The original main loop processes the data stream in the way:
```
while(length > 2)
{
   length -= 4
}
```

There will be an extra iteration when length is 4n-3, e.g. 7. consuming an extra null terminator, and it also leaves the fact that the trailing string length can only 0/1/2, so it can be processed by the following `if` statement.

While in the updated case, since the main loop is operating on a larger granularity, the same trick might violate the memory beyond the null terminator. So I used `while` here.

jkotas replied:
Can the condition in the main while loop be `while (length >= 8)` and the check `if (length >= 4)`?

I understand that the existing code does tricks with the null terminator to save a few instructions. You do not have to match those tricks.

Ruihan-Yin replied:
Sure, thanks for the suggestion, I can try that.

Ruihan-Yin committed:
update the assert check.

Ruihan-Yin commented:
The failures have been fixed. 

I would post the perf numbers I have locally. I understand there are some discussion on the implementation right now, but I would suppose it would be at worst better than the current one, so I would say the number could be reasonable.

## Bare perf numbers with the method:
 
<img width="453" alt="bareMethodBenchmark" src="https://github.com/dotnet/runtime/assets/107431934/6ba34e59-d1d3-482c-bb5e-1f6775dc0128">

Note:
The regression on the scalar path should be expected, as the vector path will introduce branching overhead, and the scalar path will get penalty for bad speculation. Other than this, numbers on the vector path look good overall. 

## Micros with ConcurrentDisctionary

String length: 1~50
<img width="802" alt="concurrentDictionary(1~50)" src="https://github.com/dotnet/runtime/assets/107431934/65510330-fea4-477b-8137-fd42ab1ff837">

String length: 20~200
<img width="799" alt="concurrentDictionary(20~200)" src="https://github.com/dotnet/runtime/assets/107431934/b1d9bc99-153b-4d99-bb06-f473171641e7">

Notes:
the default micors [ConcurrentDisctionary](https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/libraries/System.Collections/Contains) uses a dictionary with size of 512 and input string with length from 1 to 50, I also tested with strings with length from 20 to 200

Ruihan-Yin commented:
Benchmark code used for the method:

```C#
using System.Diagnostics;
using System.Numerics;
using System.Runtime.Intrinsics;
using System.Runtime;
using BenchmarkDotNet;
using BenchmarkDotNet.Running;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Toolchains.CoreRun;
using BenchmarkDotNet.Jobs;

public unsafe class BingSNR
{
    [Params(
        8,          // scalar-path
        32,         // vector-path
        64,
        128,
        512,
        10000
    )]
    public int Size;

    private byte[] _bytes, _sameBytes, _bytesDifferentCase;
    private char[] _characters, _sameCharacters, _charactersDifferentCase;

    [GlobalSetup]
    public void Setup()
    {
        _bytes = new byte[Size];
        _bytesDifferentCase = new byte[Size];

        for (int i = 0; i < Size; i++)
        {
            // let ToLower and ToUpper perform the same amount of work
            _bytes[i] = i % 2 == 0 ? (byte)'a' : (byte)'A';
            _bytesDifferentCase[i] = i % 2 == 0 ? (byte)'A' : (byte)'a';
        }
        _sameBytes = _bytes.ToArray();
        _characters = _bytes.Select(b => (char)b).ToArray();
        _sameCharacters = _characters.ToArray();
        _charactersDifferentCase = _bytesDifferentCase.Select(b => (char)b).ToArray();
    }

    [Benchmark]
    public unsafe int Hash_Vector()
    {
        fixed(char* pStr = _characters)
        {
            int ret = GetNonRandomizedHashCode_Vector(pStr, Size);
            return ret;
        }
    }
    
    [Benchmark]
    public int Hash()
    {
        fixed(char* pStr = _characters)
        {
            int ret = GetNonRandomizedHashCode(pStr, Size);
            return ret;
        }
    }

    [Benchmark]
    public unsafe int HashIgnoreCase_Vector()
    {
        fixed(char* pStr = _characters)
        {
            int ret = GetNonRandomizedHashCodeIgnoreCase_Vector(pStr, Size);
            return ret;
        }
    }


    [Benchmark]
    public int HashIgnoreCase()
    {
        fixed(char* pStr = _characters)
        {
            int ret = GetNonRandomizedHashCodeIgnoreCase(pStr, Size);
            return ret;
        }
    }

    static unsafe int GetNonRandomizedHashCode_Vector(char* src, int length)
    {
        uint* ptr = (uint*)src;
        uint hash1 = (5381 << 16) + 5381;
        uint hash2 = hash1;
        uint hash3 = hash1;
        uint hash4 = hash1;

        if((Vector128.IsHardwareAccelerated && length >= 2 * Vector128<ushort>.Count))
        {
            Vector128<uint> hashVector = Vector128.Create(hash1);

            while (length > 8)
            {
                Vector128<uint> srcVec = Vector128.Load(ptr);
                length -= 8;
                hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ srcVec;
                ptr += 4;
            }

            uint hashed1 = hashVector.GetElement(0);
            uint hashed2 = hashVector.GetElement(1);
            uint hashed3 = hashVector.GetElement(2);
            uint hashed4 = hashVector.GetElement(3);

            while (length > 4)
            {
                uint p0 = ptr[0];
                uint p1 = ptr[1];

                length -= 4;
                hashed3 = (BitOperations.RotateLeft(hashed3, 5) + hashed3) ^ (p0);
                hashed4 = (BitOperations.RotateLeft(hashed4, 5) + hashed4) ^ (p1);
                ptr += 2;
            }

            while (length > 0)
            {
                uint p0 = ptr[0];

                length -= 2;
                hashed4 = (BitOperations.RotateLeft(hashed4, 5) + hashed4) ^ (p0);
                ptr += 1;
            }

            uint res = (((BitOperations.RotateLeft(hashed1, 5) + hashed1)) ^ hashed3) + 1566083941 * (((BitOperations.RotateLeft(hashed2, 5) + hashed2)) ^ hashed4);
            return (int)res;
        }

        while (length > 8)
        {
            uint p0 = ptr[0];
            uint p1 = ptr[1];
            uint p2 = ptr[2];
            uint p3 = ptr[3];
            length -= 8;
            // hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ srcVec;
            hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (p0);
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p1);
            hash3 = (BitOperations.RotateLeft(hash3, 5) + hash3) ^ (p2);
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p3);
            ptr += 4;
        }

        while (length > 4)
        {
            uint p0 = ptr[0];
            uint p1 = ptr[1];
            length -= 4;
            // Where length is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
            hash3 = (BitOperations.RotateLeft(hash3, 5) + hash3) ^ (p0);
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p1);
            ptr += 2;
        }

        while (length > 0)
        {
            uint p0 = ptr[0];

            length -= 2;
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p0);
            ptr += 1;
        }

        uint resOnScalarPath = (((BitOperations.RotateLeft(hash1, 5) + hash1)) ^ hash3) + 1566083941 * (((BitOperations.RotateLeft(hash2, 5) + hash2)) ^ hash4);
        return (int)resOnScalarPath;
    }

    static unsafe int GetNonRandomizedHashCodeIgnoreCase_Vector(char* src, int length)
    {
        uint* ptr = (uint*)src;
        uint hash1 = (5381 << 16) + 5381;
        uint hash2 = hash1;
        uint hash3 = hash1;
        uint hash4 = hash1;
        const uint NormalizeToLowercase = 0x0020_0020u;

        if((Vector128.IsHardwareAccelerated && length >= 2 * Vector128<ushort>.Count))
        {
            Vector128<uint> hashVector = Vector128.Create(hash1);
            Vector128<uint> NormalizeToLowercaseVec = Vector128.Create(NormalizeToLowercase);
            while (length > 8)
            {
                Vector128<uint> srcVec = Vector128.Load(ptr);
                length -= 8;
                hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ (srcVec | NormalizeToLowercaseVec);
                ptr += 4;
            }

            uint hashed1 = hashVector.GetElement(0);
            uint hashed2 = hashVector.GetElement(1);
            uint hashed3 = hashVector.GetElement(2);
            uint hashed4 = hashVector.GetElement(3);

            while (length > 4)
            {
                uint p0 = ptr[0];
                uint p1 = ptr[1];

                length -= 4;
                hashed3 = (BitOperations.RotateLeft(hashed3, 5) + hashed3) ^ (p0 | NormalizeToLowercase);
                hashed4 = (BitOperations.RotateLeft(hashed4, 5) + hashed4) ^ (p1 | NormalizeToLowercase);
                ptr += 2;
            }

            while (length > 0)
            {
                uint p0 = ptr[0];

                length -= 2;
                hashed4 = (BitOperations.RotateLeft(hashed4, 5) + hashed4) ^ (p0 | NormalizeToLowercase);
                ptr += 1;
            }

            uint res = (((BitOperations.RotateLeft(hashed1, 5) + hashed1)) ^ hashed3) + 1566083941 * (((BitOperations.RotateLeft(hashed2, 5) + hashed2)) ^ hashed4);
            return (int)res;
        }

        while (length > 8)
        {
            uint p0 = ptr[0];
            uint p1 = ptr[1];
            uint p2 = ptr[2];
            uint p3 = ptr[3];
            length -= 8;
            // hashVector = (hashVector + RotateLeft(hashVector, 5)) ^ srcVec;
            hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (p0 | NormalizeToLowercase);
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p1 | NormalizeToLowercase);
            hash3 = (BitOperations.RotateLeft(hash3, 5) + hash3) ^ (p2 | NormalizeToLowercase);
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p3 | NormalizeToLowercase);
            ptr += 4;
        }

        while (length > 4)
        {
            uint p0 = ptr[0];
            uint p1 = ptr[1];
            length -= 4;
            // Where length is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
            hash3 = (BitOperations.RotateLeft(hash3, 5) + hash3) ^ (p0 | NormalizeToLowercase);
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p1 | NormalizeToLowercase);
            ptr += 2;
        }

        while (length > 0)
        {
            uint p0 = ptr[0];

            length -= 2;
            hash4 = (BitOperations.RotateLeft(hash4, 5) + hash4) ^ (p0 | NormalizeToLowercase);
            ptr += 1;
        }

        uint resOnScalarPath = (((BitOperations.RotateLeft(hash1, 5) + hash1)) ^ hash3) + 1566083941 * (((BitOperations.RotateLeft(hash2, 5) + hash2)) ^ hash4);
        return (int)resOnScalarPath;
    }

    static Vector128<uint> RotateLeft(Vector128<uint> src, int control)
    {
        return Vector128.BitwiseOr(Vector128.ShiftLeft(src, control), Vector128.ShiftRightLogical(src, 32 - control));
    }
    static unsafe int GetNonRandomizedHashCode(char* src, int length)
    {
        uint hash1 = (5381 << 16) + 5381;
        uint hash2 = hash1;
        uint* ptr = (uint*)src;
        while (length > 2)
        {
            length -= 4;
            hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ ptr[0];
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ ptr[1];
            ptr += 2;
        }
        if (length > 0)
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ ptr[0];
        return (int)(hash1 + (hash2 * 1566083941));
    }

    static unsafe int GetNonRandomizedHashCodeIgnoreCase(char* src, int length)
    {
        uint hash1 = (5381 << 16) + 5381;
        uint hash2 = hash1;
        const uint NormalizeToLowercase = 0x0020_0020u;

        uint* ptr = (uint*)src;
        while (length > 2)
        {
            length -= 4;
            hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (ptr[0] | NormalizeToLowercase);
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[1] | NormalizeToLowercase);
            ptr += 2;
        }
        if (length > 0)
            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[0] | NormalizeToLowercase);
        return (int)(hash1 + (hash2 * 1566083941));
    }

    static void Main(string[] args)
    {
        var switcher = new BenchmarkSwitcher(new[] {
        typeof(BingSNR)
      });
      switcher.Run(args);
    }
}





```

Ruihan-Yin committed:
Resolve comments.

jkotas commented:
> perf numbers 

So the break-even point is around 32 characters, and it is a regression for anything smaller than that? What can be done to address the regression for smaller strings? Also, do we know what is the string length distribution in the Bing scenarios that you are trying to improve?

Ruihan-Yin commented:
> So the break-even point is around 32 characters, and it is a regression for anything smaller than that? What can be done to address the regression for smaller strings? 

The regression mostly comes from the branch misprediction, in this method, the string length is random, makes it hard for branch predictor to learn any pattern, I was not able to come up with good ideas to fix this.

>Also, do we know what is the string length distribution in the Bing scenarios that you are trying to improve?

We did collected the distribution directly from the app, and the benchmark numbers, but I am not sure if it is proper to sure here, consider those are close-source.

jkotas commented:
> The regression mostly comes from the branch misprediction

Where are the branch mispredictions for the bare perf numbers that show the worst regressions? The benchmarks are setup to run on the same length string in a loop. There should not be any branch miserections.

Also, I have notice that your bare perf numbers are for `char[]`, but the actual use case for `string`. `char[]` and `string` payloads have different typical alignment.

Ruihan-Yin commented:
> Where are the branch mispredictions for the bare perf numbers that show the worst regressions? The benchmarks are setup to run on the same length string in a loop. There should not be any branch mispredictions.

I should made the wrong conclusion on the bare perf numbers, yes, you are right, all the input are in the same length, misprediction is not the issue in this case, I was using the trace from `Dictionary` micros

Attached is the trace with the posed micros:
<img width="1213" alt="image" src="https://github.com/dotnet/runtime/assets/107431934/667bd8ef-ab3b-40ff-8756-405ec6a5b16a">

With the highlighted line being the one taking the longest time, I suppose it might be because the folding logic for 4 elements is more complex than the original folding logic which only needs to take care of 2 elements.

> Also, I have notice that your bare perf numbers are for `char[]`, but the actual use case for `string`. `char[]` and `string` payloads have different typical alignment.

Thanks for pointing out, attached is the updated performance number:
<img width="454" alt="image" src="https://github.com/dotnet/runtime/assets/107431934/d629294d-6297-4d3c-a74e-1c414c0d3b1d">

this is the updated code:

```C#
     [Params(
        8,          // scalar-path
        32,         // vector-path
        64,
        128,
        512,
        10000
    )]
    public int Size;

    private string _string;

    [GlobalSetup]
    public void Setup()
    {
        _string = "";
        for (int i = 0; i < Size; i++)
        {
            _string += i % 2 == 0 ? 'a' : 'A';
        }        
    }

    [Benchmark]
    public unsafe int Hash_Vector()
    {
        fixed(char* pStr = _string)
        {
            int ret = GetNonRandomizedHashCode_Vector(pStr, Size);
            return ret;
        }
    }
    
    [Benchmark]
    public unsafe int Hash()
    {
        fixed(char* pStr = _string)
        {
            int ret = GetNonRandomizedHashCode(pStr, Size);
            return ret;
        }
    }

    [Benchmark]
    public unsafe int HashIgnoreCase_Vector()
    {
        fixed(char* pStr = _string)
        {
            int ret = GetNonRandomizedHashCodeIgnoreCase_Vector(pStr, Size);
            return ret;
        }
    }


    [Benchmark]
    public unsafe int HashIgnoreCase()
    {
        fixed(char* pStr = _string)
        {
            int ret = GetNonRandomizedHashCodeIgnoreCase(pStr, Size);
            return ret;
        }
    }

```

jkotas commented:
- It is not clear what the quality of the hashcodes produced by the new vectorized algorithm is. Have you done any analysis of the hashcode quality produced by the new algorithm? The existing `GetNonRandomizedHashCode` algorithm is not great so start with - #92556 has some data about it.

- For large strings, the new vectorized algorithm seems to be slower than some of the existing well-known non-cryptographic algorithms that we have implemented in the BCL. I have compared the algorithm implementation in this PR with XxHash3.HashToUInt64 and XxHash3 runs 30% faster for string length 512. Are you able to replicate this result? I think we should consider using the well-known non-cryptographic hash algorithms instead of inventing our own to address this and previous points. We cannot take a dependency on System.IO.Hashing from CoreLib for XxHash3, but we should be able to included ifdefed version of XxHash3 implementation in CoreLib to reuse some of the code - #85206 is related to this, and it has some suggestions for how to further optimize XxHash3 for this case.

- For small strings, there seems to be significant performance regression. This can be mitigated by using the existing algorithm for strings up to certain size.

jkotas commented:
> do you mean the original one? I believe that would lead to mismatch in the generated hash code on scalar path and vector path.

It would not lead to mismatch as long as the existing algorithm is unconditionally used for small strings and the vectorized algorithm is unconditionally used for long strings. Something like:

```
int GetNonRandomizedHashCode()
{
    if (length > 32) // Note there is no `Vector128.IsHardwareAccelerated` check here
        return GetNonRandomizedCodeForLongString();

    ... existing code or some other implementation optimized for small sizes ...
}

int GetNonRandomizedCodeForLongString()
{
    ... the vectorized implementation ...
}
```

Note that XxHash3 uses similar pattern to avoid paying the fixed vectorization overhead for small sizes: https://github.com/dotnet/runtime/blob/995989e8b33864648080c55e31116d9818b8760c/src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs#L129-L144

Ruihan-Yin commented:
<img width="454" alt="image" src="https://github.com/dotnet/runtime/assets/107431934/74bb7838-c31f-4994-ab0f-19a1b3cb57d4">

Thanks for the explanation,

Attached above is the bare perf number with the suggested changes. The threshold for vector path is set to be 32. The regression at small inputs is mitigated.

Ruihan-Yin commented:
Hi @jkotas , trying to kindly follow up on this PR, is there any further AR on my side to move forward on this, e.g integrating the suggested changes into the library?

jkotas commented:
Do you plan to address the regressions for small strings and make the performance for large strings even better by switching to XxHash3?

> This implementation may not resolve the known collision issue.

I am worried about your current implementation making the collision issues worse. Analyzing quality of hash functions is a non-trivial field.

Ruihan-Yin committed:
Update with suggested changes from the review.

Ruihan-Yin commented:
Failures look irrelevant.

> Do you plan to address the regressions for small strings and make the performance for large strings even better by switching to XxHash3?

The suggested changes to mitigate the small strings have been pushed to the PR. On the large input side, we currently do not have the plan to further optimize it with XxHash3.
 
> I am worried about your current implementation making the collision issues worse. Analyzing quality of hash functions is a non-trivial field.

I do understand the hash code quality analysis is non-trivial, and due to my limited experience within this domain, it could be challenging for me to perform analysis and get the conclusion. 

If this is not the right time for this PR, I understand.

jkotas commented:
> I do understand the hash code quality analysis is non-trivial, and due to my limited experience within this domain, it could be challenging for me to perform analysis and get the conclusion.

Right, it is why it is best to use one of the existing algorithms that has been analyzed.

stephentoub commented:
Given the discussion and the analysis required and the lack of movement on it for several months, I'm going to close this for now. But thanks for trying to improve things here!
